Change binder_key computation. (#10)
tlswg/draft-ietf-tls-external-psk-importer: draft-ietf-tls-external-psk-importer.md

- External Identity: The identity of an EPSK.
- Imported Identity: The identity of a PSK as sent on the wire.

# Binder Key
To prevent PSK Importers from being confused with standard OOB PSKs we change the label used in the computation of the PSK binder key.
In TLS 1.3 the PSK binder key computation is defined as follows:

> 

Reviewer: Does this need `~~~` surrounding the diagram? Ditto with the other one.

                 0
                 |
                 v
       PSK ->  HKDF-Extract = Early Secret
                 |
                 +-----> Derive-Secret(., "ext binder" | "res binder", "")
                 |                     = binder_key

We replace the string "ext binder" with "imp binder".
This means the binder key is now computed as follows:

                 0
                 |
                 v
       PSK ->  HKDF-Extract = Early Secret
                 |
                 +-----> Derive-Secret(., "ext binder"
                                        | "res binder"

>                                         | "imp binder", "")

Reviewer: Super nitpicky nitpick: there are some `|`s missing in the ASCII diagram. The left vertical line got interrupted in the line wrapping.

                 |                     = binder_key

> This change should not affect any of the securtiy properties or security proofs previously established.

Reviewer: s/securtiy/security/

- External Identity: The identity of an EPSK.
- Imported Identity: The identity of a PSK as sent on the wire.

> #Binder Key

Reviewer 2: ```suggestion
# Binder Key
```

To prevent PSK Importers from being confused with standard OOB PSKs we change the label used in the computation of the PSK binder key.

> In TLS 1.3 the  PSK binder key computation is defined as follows:

Reviewer 2: ```suggestion
In TLS 1.3 the PSK binder key computation is defined as follows:
```

                 0
                 |
                 v
       PSK ->  HKDF-Extract = Early Secret
                 |
                 +-----> Derive-Secret(., "ext binder" | "res binder", "")
                 |                     = binder_key

We replace the string "ext binder" with "imp binder".
This means the binder key is now computed as follows:

                 0
                 |
                 v
       PSK ->  HKDF-Extract = Early Secret
                 |
                 +-----> Derive-Secret(., "ext binder" 
                                        | "res binder" 
                                        | "imp binder", "")
                 |                     = binder_key

> This change should not affect any of the securtiy properties or security proofs previously established.

Reviewer 2: nit: should not? Can we state this with any more confidence?

Author: If I was being sensible / brave / mad I would re-run at least the Tamarin proofs with this change, but other than that I'd be wary of saying anything braver.

Reviewer: (I also agree it's sort of weird to have a spec say something very vague like this.)

> The only change that is not a previously valid TLS 1.3 trace we mostly focus our attention on the change to the `binder_key` computation.

Reviewer 2: This sentence doesn't parse. Can you please rephrase?

Author: Ugh, this was a copy paste error. It should read
"The only change that is not a previously valid TLS 1.3 trace in the change to the `binder_key` computation. We thus mostly focus our attention there."

> The new label prevents the key being used in other contexts, including for encrypting or decrypting messages produced elsewhere.

Reviewer 2: Hmm... I'm not sure it does. There's nothing stopping someone from deriving a binder key in the same way and using it to encrypt or MAC something. Perhaps simply:

"Label differentiation prevents collisions with non-imported external keys."

Author: I guess that's true, I was thinking in terms of types, so yes you could use it to encrypt but you wouldn't get a valid result. I'll rephrase to something correct in real life :P.

> 1. Because any keys produced using the new label are independent of keys produced using the old labels a PSK binder produced using an Importer will fail to verify when checked with the old label.

Reviewer 2: This entire list seems unnecessary. Let's remove it.

Author: I'm going through the eight security guarantees in order and arguing why this change shouldn't affect them. I should probably say that at the top.

Reviewer 2: What eight properties? (Either way, I'm not sure it's needed here. Maybe in an appendix if anywhere.)

- External Identity: The identity of an EPSK.
- Imported Identity: The identity of a PSK as sent on the wire.

# Binder Key
To prevent PSK Importers from being confused with standard OOB PSKs we change the label used in the computation of the PSK binder key.
In TLS 1.3 the PSK binder key computation is defined as follows:

~~~
             0
             |
             v
   PSK ->  HKDF-Extract = Early Secret
             |
             +-----> Derive-Secret(., "ext binder" | "res binder", "")
             |                     = binder_key
~~~

We replace the string "ext binder" with "imp binder".
This means the binder key is now computed as follows:

~~~
             0
             |
             v
   PSK ->  HKDF-Extract = Early Secret
             |
             +-----> Derive-Secret(., "ext binder"
             |                      | "res binder"
             |                      | "imp binder", "")
             |                     = binder_key
~~~

> The only change that is not a previously valid TLS 1.3 trace in the change to the `binder_key` computation.

Reviewer 3: I don't see this as a change, but more of an addition.

Author: > This seems harmless, but it is predicated on a need to separate keys that I don't understand. Is there more rationale somewhere?

The underlying issue is that without this change you can't distinguish the plain OOB case from the PSK Importer case. 
For example the server could believe it was using an importer PSK, but the client could believe it was using a vanilla OOB key. Because there is no extension in the handshake signaling the use of Importer PSKs explicitly we need to agree the use of Importer keys implicitly. 

I actually think this a weakness of TLS 1.3's OOB keys in general. They should have had a way to agree security context around the establishment of PSKs, in the same way that exporter keys allow us to bind a TLS context into a higher layer protocol. 
This change, in conjunction with PR #12 resolves this issue. 
Even if you do not want to change the hash function you can use PSK importers to bind in extra security context, and the key splitting will ensure everything Just Works™ (i.e. prove that the client and server agree on the extra security context).

Reviewer 2: @jhoyla I think @martinthomson is asking to add rationale for the new label. Perhaps you could summarize this in a single sentence in the PR?

- External Identity: The identity of an EPSK.
- Imported Identity: The identity of a PSK as sent on the wire.

# Binder Key
To prevent PSK Importers from being confused with standard OOB PSKs we change the label used in the computation of the PSK binder key.
In TLS 1.3 the PSK binder key computation is defined as follows:

~~~
             0
             |
             v
   PSK ->  HKDF-Extract = Early Secret
             |
             +-----> Derive-Secret(., "ext binder" | "res binder", "")
             |                     = binder_key
~~~

We replace the string "ext binder" with "imp binder".
This means the binder key is now computed as follows:

~~~
             0
             |
             v
   PSK ->  HKDF-Extract = Early Secret
             |
             +-----> Derive-Secret(., "ext binder"
             |                      | "res binder"
             |                      | "imp binder", "")
             |                     = binder_key
~~~

> This change should not affect any of the security properties or security proofs previously established.

Reviewer 2: ```suggestion
```

> The use of a new label means that if one party believes that it is using a standard OOB PSK, but its peer believes it is using this draft then the PSK binders will not match. 

Reviewer 2: Perhaps a simplification for this and the following two sentences:

```suggestion
This new label differentiates non-imported and imported external PSKs. Specifically, a client and server will negotiate use of an external PSK if and only if (a) both endpoints import the PSK or (b) neither endpoint imports the PSK.
```

The only change that is not a previously valid TLS 1.3 trace in the change to the `binder_key` computation.

> Adding label differentiation prevents collisions with non-imported external keys, because using a disjoint contexts in an HKDF to one disjoint ensures key separation.

Reviewer 2: ```suggestion
Adding label differentiation prevents collisions with non-imported external keys, because using disjoint contexts in an HKDF to one disjoint ensures key separation.
```

> The `binder_key` is a leaf key, and thus changing its computation doesn't affect any other key.

Reviewer 2: ```suggestion
The `binder_key` is a leaf key. Therefore, changing its computation doesn't affect any other key.
```

             |                     = binder_key
~~~

This new label differentiates non-imported and imported external PSKs. Specifically, a client and server will negotiate use of an external PSK if and only if (a) both endpoints import the PSK or (b) neither endpoint imports the PSK.

> Adding label differentiation prevents collisions with non-imported external keys, because using disjoint contexts in an HKDF ensures key separation.

Reviewer 2: This sentence and the following one are redundant.
